app/services/katello/agent/dispatcher.rb

```diff
@@ -0,0 +1,57 @@
+module Katello
+  module Agent
+    class Dispatcher
+      @supported_messages = {}
+
+      def self.register_message(name, klass)
+        @supported_messages[name] = klass
+      end
+
+      register_message(:install_package, Katello::Agent::InstallPackageMessage)
+      register_message(:remove_package, Katello::Agent::RemovePackageMessage)
+      register_message(:update_package, Katello::Agent::UpdatePackageMessage)
+      register_message(:install_errata, Katello::Agent::InstallErrataMessage)
+      register_message(:install_package_group, Katello::Agent::InstallPackageGroupMessage)
+      register_message(:remove_package_group, Katello::Agent::RemovePackageGroupMessage)
+
+      def self.dispatch(message_type, host_ids, args)
+        message_class = @supported_messages[message_type]
+
+        fail("Unsupported message type") unless message_class
+
+        uuid_data = ::Katello::Host::ContentFacet.where(host_id: host_ids).pluck(:host_id, :uuid)
+        fail("Couldn't find all hosts specified") unless host_ids.size == uuid_data.size
+
+        host_data = uuid_data.map do |host_id, consumer_id|
+          args[:consumer_id] = consumer_id
```

Comment:
Is this intentional? would you want ```args[:consumer_id]``` to be the last host in the loop ?
Wouldn't be better to not have this line and say
``` ruby
message: message_class.new(**(args.merge(consumer_id: consumer_id)))
```

Reply:
I like your idea! Better to not modify the params even though it wasn't a problem here (afaik)